create type hint stub files for module torch

t-vi opened the pull request:
We have:

- This is an initial stab at creating a type stub torch/__init__.pyi .
- This is only tested on Python 3, since that's the only Python version mypy
  works on.
- So far, we only aim at doing this for torch functions and torch.Tensor.
- Quite a few methods and functions have to be typed manually. These are
  done in torch/__init__.pyi.in

For me, PyCharm (the non-paid one) didn't seem to indicate errors in the .pyi when opening and seemed to be able to get the type hint for the few functions I tried, but I don't use PyCharm for my usual PyTorch activities, so I didn't extensively try this out.

An example of a generated PYI is at this gist. https://gist.github.com/ezyang/bf9b6a5fa8827c52152858169bcb61b1

----
Original summary:

This is more a basis for discussion that a ready solution, as it does lots of funny things, and for many of them a better solution will be found, but to facilitate discussion, I'm putting this out.

After many improvements, this could contribute to fixing #7318.

We have

- This is an initial stab at creating a type stub `torch/__init__.pyi` .
- We only do this for Python 3 because we want to use type hint introspection.
- So far, we only aim at doing this for torch functions and torch.Tensor.
- We need to import the newly build torch. Thus we do this at the end of the build.
  We use os.fork to only import the module in a child process.
- We use an annotate decorator to be able to put type hints on the native python functions in a way that
      a) they're available in the usual place for Python 3
      b) we stay Python 2 compatible
- Some annotations in torch/functional.py are provided as examples, but the remaining ones still need to be done if you are OK with the decorator approach. At first glance, it seems that the annotations are also available to Python3 introspection, e.g. on Jupyter.

For me, PyCharm (the non-paid one) didn't seem to indicate errors in the .pyi when opening and seemed to be able to get the type hint for the few functions I tried, but I don't use PyCharm for my usual PyTorch activities, so I didn't extensively try this out.

An example of a generated PYI is at [this gist](https://gist.github.com/t-vi/0d0ae013072f96f50fa11fbc2287e33b).

t-vi committed:
create type hint stub files for module torch

This is more a basis for discussion that a ready solution,
as it does lots of funny things, and for many of them
a better solution will be found.

- Initial stab at creating a type torch/__init__.pyi .
- We only do this for Python 3 because we want to
  use type hint introspection.
- So far, we only aim at doing this for torch functions
  and torch.Tensor.
- We need to import the newly build torch. Thus we
  do this at the end of the build.
  We use os.fork to only import the module in a child
  process.
- We use an annotate decorator to be able to put
  type hints on the native python functions in a way that
  a) they're available in the usual place for Python 3
  b) we stay Python 2 compatible
- Some annotatons in torch/functional.py are provided
  as examples, but the remaining ones still need to be done.

This could end up fixing #7318

t-vi committed:
the future...

SsnL commented on torch/_utils.py:
```diff
@@ -266,3 +266,13 @@ def _take_tensors(tensors, size_limit):
+# annotation decorator to get annotations in a way that is compatible
+# with both Python 2 and 3
+def annotate(ret, **kwargs):
+    def dec(fun):
```
nit: use @functools.wraps

t-vi replied:
Do I need to use wraps when I don't actually replace fun?

SsnL replied:
Oh you are right! Sorry about it. Ignore my comment above.

SsnL commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,236 @@
+                'double': 'float',
+                'Generator *': 'Generator',
+                'Generator*': 'Generator',
+                'std::vector<Tensor>': 'Tuple[Tensor, ...]',
```
should this be Union of Tuple and List?

SsnL commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,236 @@
+                'Layout': 'layout',
+                'void*': 'int',  # dataptr
+                'std::string': 'str',
+                'real': 'float',
```
sometimes this is int though

t-vi replied:
So I make it a Union, too?

SsnL replied:
I think so. Same for `accreal`.

ssnl reviewed (commented):
Nice!

t-vi committed:
Improve type mapping, add manual annotations

- Add manual tensor() annotation, thank you, @elliotwite.
- Create class stubs for device etc.
- Improve type mapping. Thank you, Simon, for the review comments.

ezyang commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,249 @@
+                'std::string': 'str',
+                'real': 'Union[float, int]',
+                'accreal': 'Union[float, int]',
+                'IntegerTensor': 'Tensor',
```
nit: I wonder if it would be possible to define some aliases here, so that we can keep a little bit more of the semantic information.

ezyang commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,249 @@
+                'std::vector<Tensor>': 'Union[Tuple[Tensor, ...],List[Tensor, ...]]',
+                'TensorList': 'Union[Tuple[Tensor, ...],List[Tensor, ...]]',
+                'Storage': 'Storage',  # inaccurate (FloatStorage...)
+                'SparseTensorRef': 'Tensor',
```
@zdevito ...this type shouldn't show up in Declarations.yaml, right?

zdevito replied:
probably not, but the handling of SparseTensor was pretty weird (at least at some point)

t-vi replied:
I'll go through them and see which ones are actually used (I used to generate hints for all functons in the yaml before I moved to "import torch and see what's actually needed").

ezyang commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,249 @@
+
+def arg_to_type_hint(arg):
+    name = arg['name']
+    if name == 'from':  # keyword...
```
lol. I wonder what our binding code does today. We should probably refuse to emit declarations that have keywords. This is a reasonable workaround for now.

ezyang commented:
We'll definitely need some sort of testing strategy for this, otherwise it will bitrot faster than you can say leroy jenkins.

ezyang commented on torch/_utils.py:
```diff
@@ -266,3 +266,13 @@ def _take_tensors(tensors, size_limit):
+
+# annotation decorator to get annotations in a way that is compatible
+# with both Python 2 and 3
+def annotate(ret, **kwargs):
```
@zdevito Now here's a kind of interesting way for JIT to support type annotations in Python 2 :)

t-vi replied:
Yeah, my main concern was "if this worked, the python people would have used that for Python 2 instead of the comments", but apparently not.

ezyang replied:
Developers are not perfectly efficient! :)

apaszke replied:
Well that might not work with mypy though. It's probably based on some static analysis/abstract interpretation, so it might not register those correctly.

t-vi replied:
It would seem not that much harder to find the annotations in the AST (in contrast to Python 2 type comments...).
Happily, we have the pyi after we're done.

ezyang commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,249 @@
+needed_modules = set()
+
+
+def type_to_python(typename):
```
I know you plan to clean this up eventually, but it would be nice to have docstrings on all of these functions.

t-vi commented:
Re testing: So the type hints should ideally have two properties:
- Code satisfying the type hints will be "correct" in some way. One could try to check this using an approach similar to the one you mention, parsing the error messages. If we are off here, programs/people will have wrong ideas about what can be passed, but it isn't worse than now.

- Any correct code will satisfy the type hint restrictions. 
  I think this may be more important property - people will get upset if their correct code is marked as wrong by checkers. One testing approach I could think of is to put "known good code" (our tests? the examples/tutorials?) through a checker (mypy?) and see if it passes. The question is whether the tests are a tad too dynamic to be statically checked and where to get good test code from if they are.

ezyang commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,249 @@
+
+
+def do_gen_py(build_lib_path):
+    while '' in sys.path:
```
lol

ezyang commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,249 @@
+        decls = [d for d in decls if d['name'] != fname + '_out']
+    for decl in decls:
+        skip = 'Type' in [a['dynamic_type'] for a in decl['arguments']]
+        if not skip:
```
```
if skip:
    continue
```

? :)

ezyang commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,249 @@
+    return name + ': ' + typename + default
+
+
+def generate_type_hints(fname, decls, is_tensor=False):
```
So, is it right to say that this applies exclusively to methods, or are functions also annotated too?

t-vi replied:
Do you mean as presented to the user or as technically implemented/as the type hints suggest?
It is called for both torch.XXX (is_tensor=False) and torch.Tensor.XXX. What happens below is that the functions torch.XXX are technically static methods in _C._VariableFunctions.

ezyang commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,249 @@
+                python_returns_s = python_returns[0]
+            type_hint = "def {}({}) -> {}: ...".format(fname, python_args_s, python_returns_s)
+            numargs = len(decl['arguments'])
+            have_vararg_version = (numargs > 0 and decl['arguments'][0]['dynamic_type'] in {'IntList', 'TensorList'} and
```
@zdevito Oh man, we forgot to talk about varargs.

ezyang commented:
OK, so I briefly skimmed the generation code, and I think the best thing we can do to make this presentable is to do some Declarations.yaml cleanup, which should simplify a bit of the hacking around you had to do here. #12562 records some overall directions we want to take here, but probably the most important prereqs for this patch are:

- Centralizing TensorOptions to kwargs transformation
- Fix some types (e.g., `Generator *` and `std::vector<Tensor>`) to no longer be directly C++ types
- Rename `dynamic_type` to `type` (after having eliminated what we previously called `type`)
- Eliminate all mentions of `Type` from Declarations.yaml
- Centralized logic for determining if a method is inplace or not
- Change `method_of` to something simpler like `function` and/or `method`. Then don't have to use the "does it have a self argument" heuristic.

t-vi commented:
So what does the #12562 washlist mean for progressing here?
I'll certainly fix the immediate comments, thanks for these!
Is it fair to say that you don't object to the annotation approach using the decorator? Then I would see to sprinkling that a bit more on functional.py.
Regarding code for testing: On the bug report, @elliotwaite mentioned the documentation examples as a source for code. That seems a great idea and we do have all the docstrings, so maybe we canmake a .py from these and run mypy.

ezyang commented:
> So what does the #12562 washlist mean for progressing here?

Let me look again at the code and work out exactly what would be affected by the changes.

> Is it fair to say that you don't object to the annotation approach using the decorator? Then I would see to sprinkling that a bit more on functional.py.

Yeah, I don't personally object to it. It seems like a quite nice way to add the annotation.

> Regarding code for testing: On the bug report, @elliotwaite mentioned the documentation examples as a source for code. That seems a great idea and we do have all the docstrings, so maybe we canmake a .py from these and run mypy.

This sounds like a great idea.

ezyang commented on tools/pyi/gen_pyi.py:
```diff
@@ -0,0 +1,249 @@
+    return type_hints
+
+
+def parameters_from_signature(sig):
```
I'm pretty unsure what this function does.

ezyang commented:
These spots look like they would make Declarations.yaml refactoring harder

https://github.com/pytorch/pytorch/pull/12500/files#diff-021fb91d72b05a5c9abadf47e7489626R56
https://github.com/pytorch/pytorch/pull/12500/files#diff-021fb91d72b05a5c9abadf47e7489626R188

Don't want to hard-code _out handling and inplace handling; should be centralized.

https://github.com/pytorch/pytorch/pull/12500/files#diff-021fb91d72b05a5c9abadf47e7489626R60

Type is going away. This one's pretty harmless though, easy enough to fix once Type is killed.

https://github.com/pytorch/pytorch/pull/12500/files#diff-021fb91d72b05a5c9abadf47e7489626R65

Hard-coded handling of self. Maybe this is OK and we'll keep this in the end. CC @zdevito 

https://github.com/pytorch/pytorch/pull/12500/files#diff-021fb91d72b05a5c9abadf47e7489626R75

Hard-coded TensorOptions expansion

https://github.com/pytorch/pytorch/pull/12500/files#diff-021fb91d72b05a5c9abadf47e7489626R87

I actually have no idea what is going on with varargs.

ezyang commented:
As a whole, after reviewing this patch, I feel a bit better about landing it before the refactor happens, after cleanups happen. I don't think it will be that much more burden when we make adjustments. (Maybe this means I'm signing up to do the adjustments :) cc @zdevito @gchanan

t-vi commented:
The `_` filtering for "internal" functions does not work as is BTW, as we need annotations for `__mul__ `.

t-vi commented on setup.py:
```diff
@@ -709,10 +709,22 @@ def get_outputs(self):
+class create_pyi(distutils.command.build.build):
+    def run(self):
+        print("-- Building .pyi --")
+        if sys.version_info[0] == 3:
```
I think the inspect module only does annotations in 3. We might just grab their code though.

ezyang commented:
@zdevito and I discussed it a bit and we're OK with merging this even before the refactors (and even before tests, if necessary). So, just do normal CR cleanup, and we should be good to go.

t-vi committed:
Updates based on feedback and testing

t-vi commented:
So I managed to run mypy over the examples and can offer that as tests.
Where would you want the tests? A new test_type_hints or test_utils or somewhere else?

t-vi committed:
Merge branch 'master' into pyi

t-vi committed:
add Tensor/Storage subclasses

t-vi committed:
more updates

t-vi committed:
add blacklist

t-vi commented:
One thing I've been wondering about is whether we want to include deprecated stuff in the type hints.
Apparently, there isn't a canonical way to mark things as deprecated in stub files.
The main things are using `Tensor(...)` constructors and also deprecated functions (e.g. the docstring examples for addcmul and addcdiv use the deprecated positional arguments instead of the keyword-only ones).

t-vi committed:
Merge branch 'master' into pyi

t-vi committed:
Add test and update generation

- Add test_type_hints.py: For testing we compile the docstring
  examples into one large file and run mypy over it.
- Make the test pass by adding lots of cases to gen_pyi.py.

t-vi committed:
Typo

t-vi committed:
Merge branch 'master' into pyi

t-vi committed:
adapt to master

t-vi committed:
no typehint testing for python2

t-vi committed:
python 3.5 compat

t-vi committed:
test for mypy

t-vi commented:
So now (except for the CI), I get a type hint file that
- let's mypy check all examples for torch and torch.Tensor docstrings except the sparse tensor one (because their examples are broken, but we didn't want to interfere with @yf225 more general efforts). There is a test that tries to call mypy (but I'm not sure whether the CI has it).
- PyCharm does not indicate errors in the PYI and seems do relatively useful things

All of this is very hacky, and I have quite a few manual annotations.
I did try to sprinkle some more commentary, but will do more.

t-vi committed:
add docstr

t-vi committed:
Merge branch 'master' into pyi

ezyang committed:
Merge remote-tracking branch 'origin/master' into t-vi-pyi

ezyang commented:
I'm really sorry this fell of the wagon. I'm looking into merging this.

t-vi commented:
When I looked at this shortly before Christmas, there were adaptation needed. Let me know if I can be of help with it.

t-vi committed:
Merge branch 'pr_12500' into pyi

t-vi committed:
Merge branch 'master' into pyi

t-vi committed:
update type hint creation to latest master

t-vi commented:
So now the type hint file makes the test with all docstrings pass again.

ezyang commented:
Yeah, I think that can wait for a follow up.

ezyang commented:
@t-vi I'm going to push some CI fixes to your branch, FYI

ezyang commented:
This PR doesn't seem to work when you do a 'develop' build. I'm investigating.

ezyang committed:
Generate pyi in-place, also general formatting and commenting

- Previously, type hints didn't work for 'develop' builds,
  because the pyi was put into a generated directory and
  consequently not picked up by the Python interpreter.
  It's a lot easier, and consistent with how we generate
  other files in PyTorch, to just generate the file in place.
  I need to test if the pyi file actually gets installed.

- I figured out why mypy didn't work in the default directory;
  it's because we had a mypy.ini which specified Python 2.
  I fixed this by making our test file Python 2 syntax compatible.

- I added some more comments, and adjusted formatting to my
  taste.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
Also, we're backporting a checked in copy of the stub file for 1.0.1

ezyang commented:
@t-vi I pushed some updates to your PR, take a look. Commit message:

```
    Generate pyi in-place, also general formatting and commenting
    
    - Previously, type hints didn't work for 'develop' builds,
      because the pyi was put into a generated directory and
      consequently not picked up by the Python interpreter.
      It's a lot easier, and consistent with how we generate
      other files in PyTorch, to just generate the file in place.
      I need to test if the pyi file actually gets installed.
    
    - I figured out why mypy didn't work in the default directory;
      it's because we had a mypy.ini which specified Python 2.
      I fixed this by making our test file Python 2 syntax compatible.
    
    - I added some more comments, and adjusted formatting to my
      taste.
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>
```

Some thoughts I had on my next pass through:

* Do you think that we shouldn't generate hints for deprecated functions, classes? Don't want to encourage people to use them :)
* Maybe we can work out how to make this work on Python 2
* (maybe some more, in a rush right now)

t-vi commented:
Thanks for doing all the hard work here!

> Do you think that we shouldn't generate hints for deprecated functions, classes? Don't want to encourage people to use them :)
Right. My main reason was that the examples had them, so the tests would not pass without.
I had the impulse to hint deprecated functions (or in lieu of that searched whether there was a deprecation decorator or so), too. It probably is a good idea, but I didn't feel like I can make the decision to force everyone to change deprecated stuff in the examples. That said, I think it was only one or two things (`addmm`?) that ran into this.

> Maybe we can work out how to make this work on Python 2
The main thing I remember is that you'd have to work around the differences in the introspection, which I'm not 100% sure is worth the trouble. That said, if making Python 2 second class that way is touchy or anyone wants that bad enough to actually do it, that would be great, too.

> Why do we say _device = device? I'm guessing there is a reason you did it this way, but I can't tell
I got a mypy error for `Tensor` methods taking an argument of type `device`, I would guess that there is some funny scoping thing going on with `Tensor.device` somehow shadowing the `torch.device` type.
This could well be a bug in mypy, but I guess the workaround doesn't hurt.

t-vi commented on test/test_type_hints.py:
```diff
@@ -93,30 +100,28 @@ def get_all_examples():
+        with open(fn, "w") as f:
+            print(get_all_examples(), file=f)
+        try:
+            # TODO: I'm not sure we actually need to start a new process
```
What I ran into trouble with was to make sure mypy picks up the right PyTorch. (Which is why I did this temporary directory dance.)

t-vi replied:
My usual workflow is either
```
python3 setup.py bdist_wheel
pip3 install dist/*.whl
pyton3 test/test_abc.py
```
or
```
python3 setup.py build
PYTHONPATH=./build/lib.<something/ pyton3 test/test_abc.py
```
one of them didn't work because mypy would either not complain enough or too much. (So I do want it to fail if there is funny stuff in the examples.)

Now that *could* be for totally unrelated reasons (bug in the mypy version I used back then, command line options that I only added later, ...), so if you find that it works for you, I'm more than happy to not use kludges like the temporary directory that.

So I'd say drop it now and if I stumble over it again, we can see again.

ezyang commented:
> This could well be a bug in mypy, but I guess the workaround doesn't hurt.

It is! https://github.com/python/mypy/issues/4146

ezyang committed:
Don't assume data exists, and comment about mypy bug workaround

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
Confirmed pyi installs correctly

ezyang committed:
Don't install mypy in environments where it doesn't work

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
@t-vi We also had some discussion about the `annotate` bits of the patch. Naively, we thought that if we just annotated the Python functions with Python 2 style type comments `# type: (...) -> int` that would be sufficient for mypy or an IDE to find out about it. But clearly you didn't do this; do you know what goes wrong? Is it something like, the pyi completely overrides everything in py, so those comments don't get picked up?

ezyang commented:
I confirmed that with PyCharm, if I create a `pyi` file, it overrides the `py` file completely; any functions which are not specified in the pyi file are not eligible for autocompletion, even if it is an ordinary function in the Python module which would be picked up in the absence of the pyi file.

ezyang commented:
What I am thinking, is that instead of using `annotate` and then reflecting the type information out, we should add a `__init__.pyi.ini` file which is used as a template which we add the rest of the generated type stubs afterwards.

The downside of this approach is that we have to remember to add pyi annotations for pure Python functions we add to `__init__.py`. But this should be pretty rare, so I am not too worried about it.

t-vi commented:
> The downside of this approach is that we have to remember to add pyi annotations for pure Python functions we add to __init__.py. But this should be pretty rare, so I am not too worried about it.

Wouldn't Python 2 annotations in functional.py work after all?
The CI might catch things when they have examples but no type hints, everything currently in functional.py except `stft` and the deprecated `potrf` has one.

ezyang commented:
I'm not sure, will need to test some more.

ezyang commented:
Even without a type hint in functional, if pure Python definitions in functional.py were respected, they should show up in completions. They do not.

However, if I add the following to `__init__.pyi`

```
from functional import *
```

Now autocompletions from functional are preserved.

There is one more subtlety. If you attach  a decorator (e.g., `annotate`) to a function, this prevents PyCharm from being able to report the argument list for a function (it still knows the function exists.) I asked about this at https://intellij-support.jetbrains.com/hc/en-us/community/posts/360002754060-Autocomplete-with-arguments-for-decorated-functions

t-vi commented:
Yeah, let's drop the decorator! (If Python2 annotations work for Python3, too, all the better.)
The import is a neat trick, probably the best way to go. If you get the annotations with that, we'd just skip functional in the pyi generation.

ezyang committed:
Restructure so that it works with Python 2

This is done by deleting all of annotation based inference; instead
we'll just write those signatures out manually.

A bit of code reorging happened along the way.

I still need to add back the direct Python implementations;
incoming soon!

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Fix some errors

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
I pushed changes to make this work on Python 2. Hypothetically we are good to go.

However, while working on these changes, I realized that since the import tricks work, we'll get even more leverage if we define a stub file for `_C.pyi` itself, and then let mypy/tools figure everything else out from files themselves. This reduces the number of hand written signatures we have to write, since C extension modules (definitionally) don't have any hand-written Python functions/methods. This is especially poignant for Tensor, which has a number of methods which the current patch doesn't document.

The primary problem with this approach is that PyCharm suggests a nonsensical originating module for the identifier in question:

![image](https://user-images.githubusercontent.com/13564/51579496-29aef800-1e8f-11e9-989e-558e2953f273.png)

I think for now, I'm going to keep just `torch/__init__.pyi`. But we should consider whether or not the alternate approach is better.

ezyang committed:
Make reexporting work

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Merge remote-tracking branch 'origin/master' into t-vi-pyi

ezyang committed:
cmake update

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Well, maybe this fixes the build

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Test script fix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Bring back py2.7 filter

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Try to make the file write more robust

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Appease lint

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Try to fix it more

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Improve docblock

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Remove torch dependency entirely

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
After a lot of whacking, the type annotation script no longer requires `import torch` to generate annotations.

ezyang commented:
Let's be clear now, your diff helped a lot :) Would have been 10x more work without it.

ezyang committed:
Swizzle the cmake a little

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Fwizzle the cmake again

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Use add_dependencies instead

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
Ugh, CI failure doesn't repro locally. Time to fire up the Docker image, I guess...

ezyang committed:
Fix cmake dependency tracking

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
More buildfixes

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Handle the rest of the problems, I think

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Merge remote-tracking branch 'origin/master' into t-vi-pyi

vsiles commented:
I'm curious: why didn't you just annotate `torch/tensor.py` directly ?

ezyang commented:
@vsiles For a pretty silly reason, actually: the original patch @t-vi wrote didn't do that, and when I tried it, PyCharm shows up the module as `torch.tensor` rather than `torch` (in reality, I tried a version where I just added a type stub for `torch._C` and then let things work out; in that case, I got `torch._C`.)

This isn't set in stone, esp. if people don't mind the originating module look ugly. But that's the reason.
